### PR TITLE
fix: embedded dashboard check

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

When a guest user with an embedded dashboard token tries to access a non-embedded dashboard we allow as long as the other dashboard is in draft or the other dashboard has no datasources or at least one datasource that the user has access.

This PR fixes the logic, so that guest users have access only if they have the proper token. I also left some comments to the discussion about the `raise_for_dashboard_access`, because the logic is very convoluted, non-intuitive, and remains so for backwards compatibility reasons.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/security/manager.py [hunk 1]
```diff
@@ -2006,28 +2006,37 @@ def raise_for_dashboard_access(self, dashboard: "Dashboard") -> None:
...
+            return
+
+        # RBAC and legacy (datasource inferred) access controls.
+        if is_feature_enabled("DASHBOARD_RBAC") and dashboard.roles:
+            if dashboard.published and {role.id for role in dashboard.roles} & {
+                role.id for role in self.get_user_roles()
+            }:
                 return
+        elif (
+            # To understand why we rely on status and give access to draft dashboards
```
Comment (michael-s-molina) [inline superset/security/manager.py:2026]: Thanks for adding the references. I agree that this logic is hard to understand and only exists for backward compatibility. I added a [card](https://github.com/orgs/apache/projects/201/views/1?pane=issue&itemId=33224540) to our major versions board to revisit this in 4.0 and break compatibility in favor of a clear and more restrictive logic.

**Review by michael-s-molina — APPROVED** [review]
LGTM

File: superset/security/manager.py (hunk 1 above)
Comment (john-bodley) [reply superset/security/manager.py:2026]: I second @michael-s-molina's comment. I think the security logic for dashboard access is rather convoluted and difficult to grok which is definitely undesirable from a security perspective, i.e., there likely are holes which can be exploited.

I wonder if in 4.0 if dashboard RBAC should no longer be a feature but rather the default which will likely simplify things even if (per @michael-s-molina's comment) this breaks compatibility.

Comment (michael-s-molina) [reply superset/security/manager.py:2026]: > I wonder if in 4.0 if dashboard RBAC should no longer be a feature but rather the default which will likely simplify things even if (per @michael-s-molina's comment) this breaks compatibility.

Interesting. I'll add this comment to the card to help the discussion.

**Pull request merged**